src/auto-compounder/AutoCompounderViews.sol

```diff
@@ -0,0 +1,114 @@
+
+        // Check that pool is initially balanced.
+        // Prevents sandwiching attacks when swapping and/or adding liquidity.
+        if (autoCompounder.isPoolUnbalanced(position)) return false;
+
+        // Get fee amounts
+        IAutoCompounder.Fees memory fees;
+        (fees.amount0, fees.amount1) = UniswapV3Logic._getFeeAmounts(id);
+
+        // Total value of the fees must be greater than the threshold.
+        if (autoCompounder.isBelowThreshold(position, fees)) return false;
+
+        // Remove initiator reward from fees, these will be send to the initiator.
+        uint256 initiatorShare = autoCompounder.INITIATOR_SHARE();
+        fees.amount0 -= fees.amount0.mulDivDown(initiatorShare, 1e18);
+        fees.amount1 -= fees.amount1.mulDivDown(initiatorShare, 1e18);
+
+        // Calculate fee amounts to match ratios of current pool tick relative to ticks of the position.
+        // Pool should still be balanced after the swap.
+        (bool zeroToOne, uint256 amountOut) = autoCompounder.getSwapParameters(position, fees);
+        bool isPoolUnbalanced = _quote(position, zeroToOne, amountOut);
+
+        isCompoundable_ = !isPoolUnbalanced;
+    }
+
+    /**
+     * @notice Off-chain view function to get the quote of a swap.
+     * @param position Struct with the position data.
+     * @param zeroToOne Bool indicating if token0 has to be swapped to token1 or opposite.
+     * @param amountOut The amount that of tokenOut that must be swapped to.
+     * @return isPoolUnbalanced Bool indicating if the pool is unbalanced due to slippage after the swap.
+     * @dev While this function does not persist state changes, it cannot be declared as view function,
+     * since quoteExactInputSingle() of Uniswap's Quoter02.sol uses a try - except pattern where it first
+     * does the swap (with state changes), next it reverts (state changes are not persisted) and information about
+     * the final state is passed via the error message in the expect.
+     */
+    function _quote(IAutoCompounder.PositionState memory position, bool zeroToOne, uint256 amountOut)
+        internal
+        returns (bool isPoolUnbalanced)
+    {
+        // Don't get quote for swaps with zero amount.
+        if (amountOut == 0) return false;
+
+        // Max slippage: Pool should still be balanced after the swap.
+        uint256 sqrtPriceLimitX96 = zeroToOne ? position.lowerBoundSqrtPriceX96 : position.upperBoundSqrtPriceX96;
+
+        // Quote the swap.
+        (, uint160 sqrtPriceX96After,,) = UniswapV3Logic.QUOTER.quoteExactOutputSingle(
+            QuoteExactOutputSingleParams({
+                tokenIn: zeroToOne ? position.token0 : position.token1,
+                tokenOut: zeroToOne ? position.token1 : position.token0,
+                amountOut: amountOut,
+                fee: position.fee,
+                sqrtPriceLimitX96: uint160(sqrtPriceLimitX96)
+            })
+        );
+
```

Comment:
Think we now don't catch the case when slippage would be very big and leftOver amountIn is smaller than required to increase the liquidity.